### PR TITLE
Make font atlas grayscale

### PR DESCRIPTION
**Pull request by floppyhammer**
# Objective

Making font atlas a grayscale texture.

## Solution

The UI pipeline is used to draw both text and sprite, so an extra uniform buffer is needed to pass a flag to differentiate between the two cases.

### CODE REVIEW

**Comment by floppyhammer** [comment]
Hello everyone! I'm new to Bevy development and I'm trying to dive in. I've been trying to get this to work, but in vain. Can anyone give me a hand? Thanks in advance.

The error says:
```
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 1, Vulkan)>`
    In a set_bind_group command
      note: bind group = `ui_params_bind_group`
    number of dynamic offsets (0) doesn't match the number of dynamic bindings in the bind group layout (1)
```

**Comment by alice-i-cecile** [comment]
I'm not skilled enough here to give you help with the debugging, but some questions on the design:

1. Why branch the UI pipeline, rather than splitting them?
2. Why grayscale, rather than binary black and white? I don't think I've ever seen a case where the letters were deliberately multi-toned.

For 1, it feels like this may be simpler, and would enable further optimization later. There's no reason why these should be rendered similarly.

For 2, I actually think I know the answer: because of antialiasing.

**Comment by floppyhammer** [comment]
@alice-i-cecile You're right about the second question. As for the first one, I guess a unified pipeline for atlas-related rendering is more desirable. It reduces code duplication as well.

**Comment by floppyhammer** [comment]
Okay, I just need to set `has_dynamic_offset` to false when creating the bind group layout.

**Comment by floppyhammer** [comment]
It's good to go. But I'm not sure if I did everything the Bevy way.

File: crates/bevy_ui/src/render/mod.rs [hunk 1]
```diff
@@ -383,13 +388,17 @@ struct UiVertex {
 pub struct UiMeta {
     vertices: BufferVec<UiVertex>,
     view_bind_group: Option<BindGroup>,
+    sprite_params: Option<Arc<UiParamsUniforms>>,
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/mod.rs:391]: Why do we need to use an `Arc` here?

File: crates/bevy_ui/src/render/mod.rs [hunk 2]
```diff
@@ -554,6 +601,11 @@ pub struct UiImageBindGroups {
     pub values: HashMap<Handle<Image>, BindGroup>,
 }
 
+#[derive(Resource, Default)]
+pub struct UiParamsBindGroups {
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/mod.rs:605]: Basic docs please: this is pub.

File: crates/bevy_ui/src/render/pipeline.rs [hunk 3]
```diff
@@ -6,10 +6,24 @@ use bevy_render::{
     view::{ViewTarget, ViewUniform},
 };
 
+#[derive(Clone, ShaderType, Hash, Eq, PartialEq)]
+pub struct UiParamsUniform {
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/pipeline.rs:10]: Definitely needs docs, including on these fields.

**Comment by floppyhammer** [comment]
@alice-i-cecile Thanks for the review. Is it any better now?

**Review by alice-i-cecile — APPROVED** [review]
This LGTM on code quality / docs / design goals.

Still needs sign off from rendering experts: I'm unsure of the architecture here.

**Comment by mockersf** [comment]
This does not handle the 2d rendering of text. You can check example `text2d`

**Comment by floppyhammer** [comment]
I start to think that adding a macro in the shader and using a dedicated pipeline for text may be a better idea. Fewer code changes this way.

**Comment by floppyhammer** [comment]
Superseded by https://github.com/bevyengine/bevy/pull/6858.

**Pull request closed without merging**